src/awkward/_v2/contents/unionarray.py

```diff
@@ -81,6 +81,13 @@ def contents(self):
     def nplike(self):
         return self._tags.nplike
 
+    def keys(self):
+        keys = set()
+        for i in range(len(self._contents)):
+            if hasattr(self._contents[i], "keys"):
+                keys.update(self._contents[i].keys)
+        return keys
+
```

Comment:
For this to work in general, all `Content` types need to have `keys` defined, and they need to be consistent about whether it's a method or a property. This implementation assumes that `self._contents[i]` has a `keys` property, which won't work if `self._contents[i]` is a UnionArray because you just defined this `keys` as a method. (No parentheses.) I think RecordArray has a `keys` property, so that will require all `Content` subclasses to have a `keys` property, not method.

The output should be a list. I don't think it was a set in the past—that would cause problems with Python 2.7 and 3.5, which don't have stable set order. I also think the `keys` of a Union _ought_ to be the intersection, rather than the union (as above), since the only keys one could pass to `_getitem_field` or `_getitem_fields` are the ones that exist on all `contents` of the UnionArray. Maybe it wasn't originally designed that way (in v1), but we have more experience with UnionArrays now: they're only useful in operations that assume the least common denominator of the `contents`, which would be an intersection here. Any test instances that require the output to be the union can be changed or deleted.

Independent of the above, this is a method that can be (and therefore should be) implemented on the Form subclasses, rather than the Content subclasses. What it returns only depends on the metadata, which is in the Form. By defining it on Forms, VirtualArray will be able to use it without unnecessarily reading data. That makes this method like `minmax_depth`, `purelist_depth`, and a few others that I implemented on the Forms a week or two ago. The Content implementations just link to the Form implementations, using a Python trick. This `keys` property should eventually become like that, but if you want to just delete the test instances that need it and move on, we can do it later.